File: day03.py
"""Solution to day 3 of Advent of Code"""

from get_input import get_input, line_parser

from common import Point


def count_trees(lines, slope):
    pos = Point(0, 0)
    total = 0
    while pos.y < len(lines):
        if lines[pos.y][pos.x % len(lines[pos.y])] == '#':
            total += 1
        pos += slope
    return total


def part1(lines):
    return count_trees(lines, Point(3, 1))


def part2(lines):
    slopes = (
        Point(1, 1),
        Point(3, 1),
        Point(5, 1),
        Point(7, 1),
        Point(1, 2)
    )
    total = 1
    for slope in slopes:
        total *= count_trees(lines, slope)
    return total


if __name__ == "__main__":
    LINES = line_parser(get_input(day=3, year=2020), parse=list)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day03.rs
use common::get_input;
use common::point::Point;

#[derive(PartialEq)]
enum Space {
    Tree,
    Open
}

type Input = Vec<Vec<Space>>;

fn count_trees(input: &Vec<Vec<Space>>, slope: &Point) -> usize {
    let mut pos = Point::new(0, 0);
    let mut total = 0;
    while (pos.y as usize) < input.len() {
        let y = pos.y as usize;
        let x = (pos.x as usize) % (input[y].len());
        if input[y][x] == Space::Tree {
            total += 1;
        }
        pos = pos + *slope;
    }
    total
}

fn part1(input: &Input) -> usize {
    count_trees(input, &Point::new(3, 1))
}

fn part2(input: &Input) -> usize {
    let slopes = vec![
        Point::new(1, 1),
        Point::new(3, 1),
        Point::new(5, 1),
        Point::new(7, 1),
        Point::new(1, 2)
    ];
    let mut total = 1;
    for slope in slopes {
        total *= count_trees(input, &slope);
    }
    total
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|line| {
            line.chars()
                .map(|c| if c == '#' { Space::Tree } else { Space::Open })
                .collect()
        })
        .collect()
}

fn main() {
    let input = parse(get_input(03, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

